src/allocator.rs

```diff
@@ -1318,20 +1309,32 @@ mod tests {
         assert_eq!(a.atom(sub).as_ref(), b"");
 
         assert_eq!(
-            a.new_substr(atom, 1, 0).unwrap_err().1,
-            "substr invalid bounds"
+            a.new_substr(atom, 1, 0).unwrap_err(),
+            EvalErr::from(AllocatorError::StartGreaterThanEnd(atom, 1, 0))
         );
         assert_eq!(
-            a.new_substr(atom, 3, 3).unwrap_err().1,
-            "substr start out of bounds"
+            a.new_substr(atom, 3, 3).unwrap_err(),
+            EvalErr::from(AllocatorError::StartOutOfBounds(
+                NodePtr::new(ObjectType::SmallAtom, 24960),
+                3,
+                2
+            ))
         );
         assert_eq!(
-            a.new_substr(atom, 0, 3).unwrap_err().1,
-            "substr end out of bounds"
+            a.new_substr(atom, 0, 3).unwrap_err(),
+            EvalErr::from(AllocatorError::EndOutOfBounds(
+                NodePtr::new(ObjectType::SmallAtom, 24960),
+                3,
+                2
+            ))
         );
         assert_eq!(
-            a.new_substr(atom, u32::MAX, 2).unwrap_err().1,
-            "substr start out of bounds"
+            a.new_substr(atom, u32::MAX, 2).unwrap_err(),
+            EvalErr::from(AllocatorError::StartOutOfBounds(
+                NodePtr::new(ObjectType::SmallAtom, 24960),
+                u32::MAX,
+                2
+            ))
```

Comment:
you can use the `matches!()` macro to check if a value matches the error code, without having to specify all the (irrelevant) details. e.g. `matches!(e, EvalErr(AllocatorError::StartOutOfBound(_, u32::MAX, 2)))`

Reply:
oh wait i see what you mean, thanks thats 10x easier.